Source/Orts.Simulation/Simulation/RollingStocks/MSTSLocomotive.cs

```diff
@@ -3016,13 +3020,54 @@ public virtual void UpdateFrictionCoefficient(float elapsedClockSeconds)
+
+                float adhesionMultiplier = (float) Math.Exp(expAdhesion * WheelSlipTimeS);
+                CurrentWheelSlipAdhesionMultiplier = adhesionMultiplier;
+
+                BaseFrictionCoefficientFactor *= adhesionMultiplier;  // Descrease friction to take into account dynamic (kinetic) friction, typically kinetic friction is approximately 50% of static friction.
+                SlipFrictionCoefficientFactor = BaseFrictionCoefficientFactor;
+
+                BaseFrictionCoefficientFactor = MathHelper.Clamp(BaseFrictionCoefficientFactor, 0.05f, 1.0f); // Ensure friction coefficient never exceeds a "reasonable" value
             }
-            else if (WheelSlip && ThrottlePercent < 0.1f && BrakeSkid) // Test to see if loco wheel is skidding due to brake application
+            else
             {
-                BaseFrictionCoefficientFactor = 0.15f * SandingFrictionCoefficientFactor;  // Descrease friction to take into account dynamic (kinetic) friction U = 0.0525
+                WheelSlipTimeS = 0; // Reset slip time if wheel slip stops
+
+                if (SlipFrictionCoefficientFactor < BaseFrictionCoefficientFactor && SlipFrictionCoefficientFactor != 0) // Once these two are equal then assume that wheels have stopped slipping.
+                {
+                    //                    Trace.TraceInformation("SlipFriction {0} Base {1}", SlipFrictionCoefficientFactor, BaseFrictionCoefficientFactor);
+                    // Exponential curve is used to transition between dynamic friction and static friction when wheel stops slipping
+                    // Constant points facilitate an increase from 0.7 to 1 in 3 seconds - P1 = (5, 0.77), P2 = (0, 1). Hence exp constant = 0.0523
+                    var expAdhesion = 0.0523;
+                    WheelStopSlipTimeS += elapsedClockSeconds;
+                    WheelStopSlipTimeS = MathHelper.Clamp(WheelStopSlipTimeS, 0.0f, 5.0f); // Ensure that time to transition between the two friction cases is maintained - currently set to 3 secs
+
+                    float adhesionMultiplier = CurrentWheelSlipAdhesionMultiplier * (float)Math.Exp(expAdhesion * WheelStopSlipTimeS);
+
+//                    Trace.TraceInformation("adhesion {0} StopTime {1} Base {2} Current {3}", adhesionMultiplier, WheelStopSlipTimeS, BaseFrictionCoefficientFactor, CurrentWheelSlipAdhesionMultiplier);
+
+                    BaseFrictionCoefficientFactor *= adhesionMultiplier;  // Descrease friction to take into account dynamic (kinetic) friction, typically kinetic friction is approximately 50% of static friction.
+                    SlipFrictionCoefficientFactor = BaseFrictionCoefficientFactor;
+                    BaseFrictionCoefficientFactor = MathHelper.Clamp(BaseFrictionCoefficientFactor, 0.05f, 1.0f); // Ensure friction coefficient never exceeds a "reasonable" value
+                }
```

Comment:
Peter, the base friction coefficient is also used for the axle model, which already takes into account this kind of variations, so in fact we are introducing twice the same effects. Could you move this calculations to SteamLocomotive.cs, or add a check to avoid it for diesel&electric locos?

Reply:
> the base friction coefficient is also used for the axle model, which already takes into account this kind of variations, so in fact we are introducing twice the same effects. Could you move this calculations to SteamLocomotive.cs, or add a check to avoid it for diesel&electric locos?

I have added a temporary patch to apply this only to steam locomotives (subject to further discussion).

I strongly believe that adhesion coefficients should only be calculated in one location, ie MSTSLocomotive with the outcome being passed to the axle model, or steam model as required. I think that it is confusing to be calculating adhesion in two different places, and it also runs the risk that any future code changes to the adhesion coefficient may not be altered in both locations if they remain separate.

The other aspect to be considered is that the adhesion coefficient is displayed in the HuD, is the value calculated in the Axle module being displayed in the HuD?

I would also like to see conversion factors used where possible, rather then using discrete numbers. For example, 

`speed = Math.Abs(3.6f*speed);`

should be 

`speedkmph = (MpS.ToKpH(AbsSpeedMpS);`

Given the fact that this PR and your recently opened PR (#627) have some degree of overlap, and thus are interacting with each other, and also given your stated desire to use the axle model for steam locomotives, I would suggest that we co-ordinate these activities, close one of these PRs and combine the relevant code changes.

Perhaps we can discuss this work consolidation in the PM that we currently have going.

Reply:
> I have added a temporary patch to apply this only to steam locomotives (subject to further discussion).
Thanks for that.
> I strongly believe that adhesion coefficients should only be calculated in one location, ie MSTSLocomotive with the outcome being passed to the axle model, or steam model as required. I think that it is confusing to be calculating adhesion in two different places, and it also runs the risk that any future code changes to the adhesion coefficient may not be altered in both locations if they remain separate.

I more or less agree. In my opinion the _maximum_ adhesion coefficient should be calculated in a global class (probably better in MSTSWagon or TrainCar, so wagon brakes are also affected). With maximum adhesion coefficient I mean the static coefficient, given by Curtius-Kniffler formula and corrected due to weather and sanding. The change of the adhesion factor with time or slip percentage will depend on the model being used. Ideally we should calculate also the dynamically-varying adhesion factor outside the model too, but unfortunately the formula used in Axle.cs does not allow doing so (Polach 2005 formula does need a varying adhesion factor, but as I've said it has too many free parameters).

> The other aspect to be considered is that the adhesion coefficient is displayed in the HuD, is the value calculated in the Axle module being displayed in the HuD?

I have no idea. Strictly speaking, the axle model does not calculate an effective adhesion coefficient, but it calculates the  tangential force per unit of weight, so I don't know.

>  would also like to see conversion factors used where possible, rather then using discrete numbers. For example,
> speed = Math.Abs(3.6f*speed);
> should be
> speedkmph = (MpS.ToKpH(AbsSpeedMpS);

Yes, I believe this is required by OR coding conventions. Another thing that I like when reading code (although I don't know if it is mandatory) is to have uppercase names for variables when they produce side effects outside the current method (e.g. public global variables), and lowercase for local variables created in intermediate steps.

> Given the fact that this PR and your recently opened PR (https://github.com/openrails/openrails/pull/627) have some degree of overlap, and thus are interacting with each other, and also given your stated desire to use the axle model for steam locomotives, I would suggest that we co-ordinate these activities, close one of these PRs and combine the relevant code changes.

I don't mind resolving git conflicts, so I suggest that you keep going with this PR, and I'll correct mine if there are conflicts. I have to study if the integration time in the axle model is good enough to allow the particularities of steam traction (variable force during rotation), so meanwhile we can keep the two different models.

What else do you have in mind for including in this PR?

Reply:
> I more or less agree. In my opinion the _maximum_ adhesion coefficient should be calculated in a global class (probably better in MSTSWagon or TrainCar, so wagon brakes are also affected). With maximum adhesion coefficient I mean the static coefficient, given by Curtius-Kniffler formula and corrected due to weather and sanding. 
>

I believe that wheel adhesion during braking and braking skid may be different to the adhesion for a locomotive, especially where wheel slip correction is used to increase the adhesion of the locomotive.

As long as it caters correctly for the different scenarios, and is all together, I would have no problems if it is in TrainCar, or a suitable other class.

>The change of the adhesion factor with time or slip percentage will depend on the model being used. Ideally we should calculate also the dynamically-varying adhesion factor outside the model too, but unfortunately the formula used in Axle.cs does not allow doing so (Polach 2005 formula does need a varying adhesion factor, but as I've said it has too many free parameters).
> 
What values from the SlipCharacteristics calculations cannot be bought out of the axle model, and why can't it be calculated elsewhere.

> I have no idea. Strictly speaking, the axle model does not calculate an effective adhesion coefficient, but it calculates the tangential force per unit of weight, so I don't know.
> 
Regardless, it is still a "dynamic adhesion coefficient" and as such should be visible to the user for debug purposes as the other adhesion coefficients already are.

> I don't mind resolving git conflicts, so I suggest that you keep going with this PR, and I'll correct mine if there are conflicts. I have to study if the integration time in the axle model is good enough to allow the particularities of steam traction (variable force during rotation), so meanwhile we can keep the two different models.
> 
> What else do you have in mind for including in this PR?

At the moment I am still not 100% happy that the steam model is performing in a similar fashion as video evidence suggests. Until I can explain the difference or get it to align, I want to reflect on a few aspects of the model, and do additional testing.

Reply:
> I believe that wheel adhesion during braking and braking skid may be different to the adhesion for a locomotive, especially where wheel slip correction is used to increase the adhesion of the locomotive.

I don't think so, the wheel-rail contact is the same. From a physical point of view, a tractive force applied to the wheel is the same as a brake force slowing the wheel, the only differences being on how does that force changes with time. Slip/skid control does not increase the maximum possible adhesion, but modifies applied forces so the adhesion limit is not reached.

> What values from the SlipCharacteristics calculations cannot be bought out of the axle model, and why can't it be calculated elsewhere.

The formula in axle.cs is:
![image](https://user-images.githubusercontent.com/22003819/160416520-aa11aa97-1424-4064-8299-c7507e0ee7ab.png)
Where the mu_0 parameter is the maximum (static) adhesion factor. So this formula requires a constant friction coefficient, i.e. not dependent on slip percentage.

On the other hand, for example in Polach model:
![image](https://user-images.githubusercontent.com/22003819/160416851-2841512c-a1bc-4077-b7fb-ff26d10e4de1.png)
The "mu" coefficient is dependent on the wheel speed, so this formula accepts a friction factor varying during the wheelslip process.

Both approaches are correct from my point of view, so that's why I prefer to calculate the maximum factor in TrainCar.cs, and include the dependence with wheelslip in the specific adhesion model being used.

Do you have any references where I can find some information about Curtius&Kniffler equation?

Reply:
Rather then spreading our conversation over two places, I will respond to some of this in the PM that we already have going.